File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-darkreader-inline-fill-4aa0f2-data-darkreader-inline-fillpolygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2824-count-pairs-whose-sum-is-less-than-target/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-darkreader-inline-fill-4aa0f2-data-darkreader-inline-fillpolygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2824-count-pairs-whose-sum-is-less-than-target.go
package main

import (
	"fmt"
	"sort"
)

func countPairs(array []int, target int) int {
	
    sort.Ints(array)
    
	numOfPairs := 0
	leftIndex := 0
	backIndex := len(array) - 1

	for leftIndex < backIndex {
        if (array[leftIndex]+array[backIndex] < target) {
			currNumOfPairs := backIndex - leftIndex
			numOfPairs += currNumOfPairs
			leftIndex++
		} else {
			backIndex--
		}
	}

	return numOfPairs
}

// func main() {
// 	array := []int{-6, 2

File: 0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-darkreader-inline-fill-4aa0f2-data-darkreader-inline-fillpolygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2824-count-pairs-whose-sum-is-less-than-target/0div-idbig-omega-company-tagsdiv-idbig-omega-topbardiv-classcompanytagscontainer-styleoverflow-x-scroll-flex-wrap-nowrap-div-classcompanytagscontainer-tagno-companies-found-for-this-problem-div-divdiv-classcompanytagscontainer-chevrondivsvg-version11-idicon-xmlnshttp-wwww3org-2000-svg-xmlns-xlinkhttp-wwww3org-1999-xlink-x0px-y0px-viewbox0-0-32-32-fill4087f1-xml-spacepreserve-stylewidth-20px-darkreader-inline-fill-4aa0f2-data-darkreader-inline-fillpolygon-points16-22-6-12-74-106-16-192-246-106-26-12-polygonrect-id-x3c-transparent-rectangle-x3e-classst0-fillnone-width32-height32-rect-svg-div-div-div-div2824-count-pairs-whose-sum-is-less-than-target.php
class Solution {
// <?php

function countPairs($array, $target) {
    sort($array);

    $numOfPairs = 0;
    $leftIndex = 0;
    $backIndex = count($array) - 1;
    
    while ($leftIndex < $backIndex) {
        if ($array[$leftIndex] + $array[$backIndex] < $target) {
            $currNumOfPairs = $backIndex - $leftIndex;
            $numOfPairs += $currNumOfPairs;
            $leftIndex++;
        } 
        else {
            $backIndex--;
        }
    }
    return $numOfPairs;
}
}

// $array = [-6, 2, 5, -2, -7, -1, 3];
// $target = -2;
// echo countPairs($array, $target);

// countPairs([-6, 2, 5, -2, -7, -1, 3], -2) //10